### ISSUE TITLE
Navigation crash : Any tile edge can only be used once as part of the geometry

### ISSUE THREAD

**Issue by diegolaballos** [mapbox/mapbox-navigation-android#2337]
Crash when starting navigation:

```
Fatal Exception: java.lang.Error: Any tile edge can only be used once as part of the geometry
       at com.mapbox.navigator.Navigator.getRouteBufferGeoJson(Navigator.java)
       at com.mapbox.services.android.navigation.v5.navigation.MapboxNavigator.retrieveRouteGeometryWithBuffer(MapboxNavigator.java:99)
       at com.mapbox.services.android.navigation.v5.navigation.NavigationRouteProcessor.updateRoute(NavigationRouteProcessor.java:64)
       at com.mapbox.services.android.navigation.v5.navigation.NavigationRouteProcessor.buildNewRouteProgress(NavigationRouteProcessor.java:43)
       at com.mapbox.services.android.navigation.v5.navigation.RouteProcessorRunnable.process(RouteProcessorRunnable.java:63)
       at com.mapbox.services.android.navigation.v5.navigation.RouteProcessorRunnable.run(RouteProcessorRunnable.java:46)
       at android.os.Handler.handleCallback(Handler.java:873)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:226)
       at android.os.HandlerThread.run(HandlerThread.java:65)
```

**Android API: Android 9**
**Mapbox Navigation SDK version:*0.42*

Not sure how to reproduce the crash. The only information I have is that it happens when starting navigation.

**Comment by Guardiola31337** [comment]
We should check if this could be an issue in `1.0` @mapbox/navigation-android 

BTW does that ring a 🔔 @kevinkreiser @mskurydin?

**Comment by Guardiola31337** [comment]
Any ideas @mskurydin @SiarheiFedartsou?

**Comment by diegolaballos** [comment]
We're currently migrating to 1.0 so still not sure. I'd open another ticket if it happens again there

**Comment by Guardiola31337** [comment]
@diegolaballos That makes sense 👍 

Good to close here then?

Friendly reminder that you can review the developer [documentation](https://docs.mapbox.com/android/beta/navigation/overview/) to start building with the Mapbox Navigation SDK v1.0 for Android and check out the [migration guide](https://github.com/mapbox/mapbox-navigation-android/wiki/1.0-Navigation-SDK-Migration-Guide) to transition your project from the "legacy" Navigation SDK to the 1.0 version and let us know if you have any questions or run into issues. We would love your feedback.

**Comment by zugaldia** [comment]
We've heard of another customer integrating v1.0 hitting this issue. Reopening to prioritize an investigation.

**Comment by zugaldia** [comment]
Possibly related: https://github.com/mapbox/mapbox-navigation-android/issues/1988 and https://github.com/mapbox/mapbox-navigation-android/issues/1971#issuecomment-664660165.

**Comment by mark-lapasa** [comment]
blocked on this same issue too

**Comment by zugaldia** [comment]
/cc: @etl could you have a look?

**Comment by SiarheiFedartsou** [comment]
@LukasPaczos @Guardiola31337 this issue seems to be trivial to debug if we could obtain any route for which we have this exception. Is it possible to find one?

**Comment by LukasPaczos** [comment]
@mark-lapasa would you be able to share the coordinates you are using for the route request that produces this exception?

**Comment by SiarheiFedartsou** [comment]
@mark-lapasa and not only coordinates, but any parameters you are passing to Directions API, please

**Comment by mark-lapasa** [comment]
> @mark-lapasa and not only coordinates, but any parameters you are passing to Directions API, please
> 
> **UPDATE**
> Sorry, didn't notice all route parameters from your snippet.

Hope the following helps:

```
class SandboxFragment : BaseFragment() {

    private lateinit var binding: FragmentSandboxBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentSandboxBinding.inflate(inflater, container, false).apply {
//            configureMapView(requireActivity(), mapView,  Pair(48.85819, 2.29458), "Eiffel Tower")
            configureMapView(
                requireActivity(),
                mapView,
                listOf(
                    Place(lat = 48.85819, lng = 2.29458),
                    Place(lat = 48.85667232096175, lng = 2.2923376716963206)
                ),
                viewLifecycleOwner
            )
        }

        return binding.root
    }
}

```

**Comment by mskurydin** [comment]
@mark-lapasa Do you see this exception consistently on every run? Would it be possible for you to save the contents of `routes: List<DirectionsRoute>` into a file and attach to the issue? This would help a lot! [having trouble reproducing it with a route request]

**Comment by mark-lapasa** [comment]
> @mark-lapasa Do you see this exception consistently on every run? Would it be possible for you to save the contents of `routes: List<DirectionsRoute>` into a file and attach to the issue? This would help a lot! [having trouble reproducing it with a route request]

@mskurydin This is a snapshot of what `routes` looks like. I just reproduced the issue again.

![image](https://user-images.githubusercontent.com/74388312/101681514-3a8b9700-3a30-11eb-8229-d8051d5c5648.png)

**Comment by mark-lapasa** [comment]
@mskurydin Also to add more specifics, these are exact MapBox classes used in my original `configureMapView()` [post](https://github.com/mapbox/mapbox-navigation-android/issues/2337#issuecomment-741175109):

![image](https://user-images.githubusercontent.com/74388312/101681803-bb4a9300-3a30-11eb-8429-6311f62596aa.png)

**Comment by Guardiola31337** [comment]
@mark-lapasa for your own security going to remove the screenshot you shared that included potentially your access token 😬 

Feel free to upload another screenshot hiding it or share plain text.

**Comment by mskurydin** [comment]
What are the coordinates for the route shown in the picture?
According to the encoded geometry, it shows Somalia (1 point):
lon/lat - [47.302239,3.75299],[47.302239,3.75299]

**Comment by mark-lapasa** [comment]
> @mark-lapasa for your own security going to remove the screenshot you shared that included potentially your access token 😬
> 
> Feel free to upload another screenshot hiding it or share plain text.

Thx - I was gonna refresh all the tokens when it was going to prod. Doesn't help I had committed that secret key into the repro, lol. I'm solo developer on this project.

**Comment by mark-lapasa** [comment]
@mskurydin @Guardiola31337 - Second attempt at capturing the breakpoint. Not sure how in Android studio how to dump to text file.

![image](https://user-images.githubusercontent.com/74388312/101683776-6a886980-3a33-11eb-8b29-514ec34fab44.png)

**Comment by LukasPaczos** [comment]
@mark-lapasa could you call `routes[0].toJson()` and paste the route JSON here? Please make sure to remove the access token that's appended to the end of that JSON string.

**Comment by mskurydin** [comment]
We have found the exact route responses and see that the lon/lat are confused during the request:

From the request (it's over the ocean)
```json
"coordinates": [{"lon": 48.85819, "lat": 2.29458}, {"lon": 48.856672, "lat": 2.292338}]
```

Our services follow the lon-first model most of the time.

--

In any case, we do the follow-up investigation of why we get a weird route and why it causes a crash.

**Comment by mark-lapasa** [comment]
@mskurydin yayyyyyyy! I am unblocked thanks to your tip. The problem was as you stated.

```
   // Causes crash
    val points = latLngs.map {
        Point.fromLngLat(it.latitude, it.longitude)
    }
```

```
   // Causes no crash
    val points = latLngs.map {
        Point.fromLngLat(it.longitude, it.latitude)
    }
```

I was so used to the documentation going from latitude followed with longitude that I erroneously believed it was the same for `Point.fromLngLat()`

I have achieved my expected result. Thx all!

![image](https://user-images.githubusercontent.com/74388312/101811785-8732a900-3ae8-11eb-8a0a-336e6cbc711a.png)

**Comment by Guardiola31337** [comment]
BTW does the crash force close the app? If so, we should think about how to notify the user in this case (lat, lon is reversed) instead of bubble up the exception.

cc @LukasPaczos @mskurydin

**Comment by mark-lapasa** [comment]
@Guardiola31337 In my case, it leads to an unrecoverable crash because I'm not explicitly catching the Throwable.

`java.lang.Error: Any tile edge can only be used once as part of the geometry`

**Comment by Guardiola31337** [comment]
Thanks for confirming @mark-lapasa 🙇‍♂️

**Comment by mskurydin** [comment]
@Guardiola31337 both the leaking exception and the invalid handling of a 1-point route are fixed internally. In case the lon/lat are confused and the location is far from any drivable edges, there simply won't be any route returned. Can we close the issue?

**Comment by diegolaballos** [comment]
Yes, seems it's not happening anymore for me

**Comment by Guardiola31337** [comment]
Closing as fixed 🚀
